crates/ty_python_semantic/resources/mdtest/assignment/annotations.md

```diff
@@ -495,3 +493,26 @@ def _(i: int):
     reveal_type(b)  # revealed: list[Unknown]
     reveal_type(c)  # revealed: list[Unknown]
 ```
+
+The function arguments are inferred using the type context:
+
+```py
+from typing import TypedDict
+
+class TD(TypedDict):
+    x: int
+
+def f[T](x: list[T]) -> T:
+    return x[0]
+
+a: TD = f([{ "x": 0 }, { "x": 1 }])
+reveal_type(a)  # revealed: TD
+
+b: TD | None = f([{ "x": 0 }, { "x": 1 }])
+# TODO: Narrow away the `None` here.
+reveal_type(b)  # revealed: TD | None
```

Comment:
This is really tricky. We infer the list argument to be `list[TD | None]` due to the annotated type of `T` being `TD | None`. I'm not sure there's a clear way to filter away the `None` here.